heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java

```diff
@@ -347,14 +347,18 @@ protected boolean validateRuntimeManage(
       return false;
     }
 
-    // Check whether cluster/role/environ matched
-    ExecutionEnvironment.ExecutionState executionState = adaptor.getExecutionState(topologyName);
-    if (executionState == null
-        || !executionState.getCluster().equals(Context.cluster(config))
-        || !executionState.getRole().equals(Context.role(config))
-        || !executionState.getEnviron().equals(Context.environ(config))) {
-      LOG.severe("cluster/role/environ not matched");
-      return false;
+    // Check whether cluster/role/environ matched. It is possible to get into a state where some of
+    // the data does not actually exist in zookeeper. If we are in this state them assume the
+    // runtime is valid.
+    if (adaptor.doesExecutionStateExist(topologyName)) {
```

Comment:
+1 to your first paragraph. We could limit this change to say that unless the command is a `kill` we should require valid system state. Kill should hard-kill everything and clean out all state.

I don't think the 2nd paragraph suggestion would work, since AFAIK execution state needs to persist across process failures so ephemeral nodes wouldn't work. Those are better suited for state that truly should disappear when the process dies, like tmaster location.